pydra/engine/boutiques.py

```diff
@@ -0,0 +1,221 @@
+import typing as ty
+import json
+import attr
+import subprocess as sp
+import os
+from pathlib import Path
+from functools import reduce
+
+from ..utils.messenger import AuditFlag
+from ..engine import ShellCommandTask
+from ..engine.specs import (
+    SpecInfo,
+    ShellSpec,
+    ShellOutSpec,
+    File,
+    Directory,
+    attr_fields,
+)
+from .helpers import ensure_list, execute
+from .helpers_file import template_update, is_local_file
+
+
+class BoshTask(ShellCommandTask):
+    """Shell Command Task based on the Boutiques descriptor"""
+
+    def __init__(
+        self,
+        zenodo=None,
+        bosh_file=None,
+        audit_flags: AuditFlag = AuditFlag.NONE,
+        cache_dir=None,
+        input_spec: ty.Optional[SpecInfo] = None,
+        messenger_args=None,
+        messengers=None,
+        name=None,
+        output_spec: ty.Optional[SpecInfo] = None,
+        rerun=False,
+        strip=False,
+        **kwargs,
+    ):
+        """
+        Initialize this task.
+
+        Parameters
+        ----------
+        zenodo: :obj: str
+            Zenodo ID
+        bosh_file : : str
+            json file with the boutiques descriptors
+        audit_flags : :obj:`pydra.utils.messenger.AuditFlag`
+            Auditing configuration
+        cache_dir : :obj:`os.pathlike`
+            Cache directory
+        input_spec : :obj:`pydra.engine.specs.SpecInfo`
+            Specification of inputs.
+        messenger_args :
+            TODO
+        messengers :
+            TODO
+        name : :obj:`str`
+            Name of this task.
+        output_spec : :obj:`pydra.engine.specs.BaseSpec`
+            Specification of inputs.
+        strip : :obj:`bool`
+            TODO
+
+        """
+        if (bosh_file and zenodo) or not (bosh_file or zenodo):
+            raise Exception("either bosh or zenodo has to be specified")
+        elif zenodo:
+            bosh_file = self._download_spec(zenodo)
+
+        # retry logic - an error on travis is raised randomly, not able to reproduce
+        tries, tries_max = 0, 7
+        while tries < tries_max:
+            try:
+                with bosh_file.open() as f:
+                    self.bosh_spec = json.load(f)
+                break
+            except json.decoder.JSONDecodeError:
+                tries += 1
+                if tries == tries_max:
+                    raise
+
+        if input_spec is None:
+            input_spec = self._prepare_input_spec()
+        self.input_spec = input_spec
+        if output_spec is None:
```

Comment:
you're right, haven't even tested it...

Reply:
decided to introduce `input/output_spec_names` instead of `input/output_spec`, so you can provide subset of names that should be used, but type, help_string, etc. is taken from the zenodo spec file. Otherwise I wasn't sure how to deal with conflicts in types, help_strings etc.